### DISCUSSION TITLE
Access non-local resources via SSH/SFTP

### DISCUSSION THREAD

**Discussion by gian-didom** [apple/pkl#522] [Show and tell]
From Pkl documentation, it seems that it is possible to access a non-local resources with an HTTP(S) URL.
I'm using Pkl to setup a distributed local environment, in which the machines do not have a running HTTP(S) server; instead, they all have an SSH and SFTP server.

I think it could be a great addition to use URLs such as `ssh://remote-hostname/path/to/file.pkl` or `sftp://remote-hostname/path/to/file.pkl` to automatically load remote resources without complex setups. I think that this is currently achievable by using some kind of filesystem mounting, but maybe there could be a more elegant way to achieve this.

EDIT: At https://github.com/gian-didom/pkl-ssh/tree/sftp there's a version that uses Apache MINA SSH library to add support for SFTP.
It works well; however, it could be more elegant to only provide an SFTP filesystem provider and leave everything else as-is. Also, I wasn't able to generate native images due to linking and initialization errors.

**Comment by gian-didom** [comment]
I was able to spin-up an implementation of SSH file access via `ssh://` protocol using JSch. For now, I only enable key-based access since I don't think that pushing users to use the password is a good idea. 

Do you think it I should submit a pull request?

**Reply by holzensp** [reply]
For such things to land in Pkl itself, they must be quite well-rounded (i.e. cover everything people can reasonably expect to be covered when they read `ssh:` URIs). As it seems you've already done quite a bit of work to make it concrete, I'd suggest putting up a SPICE (see [here](https://github.pie.apple.com/pkl/pkl-evolution-staging)) and inviting some feedback from the community.

Mind you, this does not strictly  _need_ to be covered in Pkl itself, because Pkl has SPIs for custom Resource Readers (see [here](https://pkl-lang.org/main/current/language-reference/index.html#extending-resource-readers) and [here](https://github.com/apple/pkl/blob/0.26.0/pkl-core/src/main/java/org/pkl/core/resource/ResourceReader.java)) and `ModuleKey`s (see [here](https://github.com/apple/pkl/blob/0.26.0/pkl-core/src/main/java/org/pkl/core/module/ModuleKey.java)). The latter (`ModuleKey`s) are what you seem to be looking for (because `import`). These SPIs are also available in our language bindings, so it need to even be in Java.

**Reply by holzensp** [reply]
Oh, and then I totally forget to mention: Cool! ;)
